mbtools/validate_mbz_html.py

```diff
@@ -306,6 +310,129 @@ def find_ib_uuid_violations(html_elements):
+
+        for td_or_th in tr:
+            if not is_table_child_allowed(td_or_th.tag):
+                msg = f"{td_or_th.tag} is not allowed"
+                violations.append(Violation(TABLE_VIOLATION + msg,
+                                            elem_location))
+
+            if td_or_th.tag == 'th':
+                violations += find_th_violations(td_or_th, element.tag,
+                                                 elem_location)
+    return violations
+
+
+def find_th_violations(td_or_th, element_tag, elem_location):
+    violations = []
+
+    if element_tag == 'thead':
+        scope = td_or_th.get('scope')
+        if scope != 'col':
+            msg = 'must include scope attribute in thead th with value col'
+            violations.append(Violation(TABLE_VIOLATION + msg, elem_location))
+
+    elif element_tag == 'tbody':
+        scope = td_or_th.get('scope')
+        if scope != 'row':
+            msg = 'must include scope attribute in tbody th with value row'
+            violations.append(Violation(TABLE_VIOLATION + msg, elem_location))
+
+    return violations
+
+
+def find_table_no_th_violation(table, elem_location):
+    violations = []
+
+    has_th_in_tbody = False
+    has_th_in_thead = False
+
+    for thead_or_tbody in table.getchildren():
+        for tr in thead_or_tbody:
+            for th_or_td in tr:
+                if th_or_td.tag == 'th' and thead_or_tbody.tag == 'tbody':
+                    has_th_in_tbody = True
+
+                if th_or_td.tag == 'th' and thead_or_tbody.tag == 'thead':
+                    has_th_in_thead = True
+
+    if table.get('class') == 'os-raise-doubleheadertable':
+        if not (has_th_in_tbody and has_th_in_thead):
+            msg = "doubleheadertable requires th in both thead and tbody"
+            violations.append(Violation(TABLE_VIOLATION + msg, elem_location))
+    else:
+        if not (has_th_in_thead or has_th_in_tbody):
+            msg = "th is required in either thead or tbody"
+            violations.append(Violation(TABLE_VIOLATION + msg, elem_location))
+
+    return violations
+
+
+def find_table_violations(html_elements):
+    violations = []
+
+    for elem in html_elements:
+        tables = elem.get_elements_by_name('table')
+        for table in tables:
+            elem_location = elem.location
+
+            violations += find_missing_class_violations(table, elem_location)
+            violations += find_invalid_children_violations(table,
+                                                           elem_location)
+            violations += find_missing_thead_tbody_violations(table,
+                                                              elem_location)
+            violations += find_table_no_th_violation(table, elem_location)
+            for element in table.getchildren():
+                violations += find_invalid_element_violations(element,
+                                                              elem_location)
```

Comment:
If we use something like the revised version of the `find_invalid_element_violations` above, this can just be a single call vs a loop:

```python
violations += find_invalid_element_violations(table, elem_location)
```